dont-wreck-my-house/src/main/java/learn/lodging/domain/ReservationService.java

```diff
@@ -0,0 +1,199 @@
+            }
+        }
+        if (existing == null){
+            result.addErrorMessage("Reservation not found.");
+            return result;
+        }
+        LocalDate start = reservation.getStartDate();
+        LocalDate end = reservation.getEndDate();
+        reservation.setTotal(calculateTotal(start, end, reservation.getHostId()));
+
+        boolean success = reservationRepository.update(reservation);
+        if (!success){
+            result.addErrorMessage("Reservation not found.");
+        }
+        return result;
+    }
+
+    public Result<Reservation> delete(Reservation reservation) throws DataException {
+        Result<Reservation> result = validate(reservation);
+        if (!result.isSuccess()){
+            return result;
+        }
+        if (reservation == null){
+            result.addErrorMessage("Reservation not found.");
+        }
+
+        boolean success = reservationRepository.delete(reservation);
+        if (!success){
+            result.addErrorMessage("Reservation not found.");
+        }
+        return result;
+    }
+
+    public BigDecimal calculateTotal (LocalDate start, LocalDate end, String hostId){
+        BigDecimal standard = hostRepository.findById(hostId).getStandardRate();
+        BigDecimal weekend = hostRepository.findById(hostId).getWeekendRate();
+        int numStandard = 0;
+        int numWeekend = 0;
+
+        for (;start.compareTo(end) <= 0; start = start.plusDays(1)){
+            if (start.getDayOfWeek() == DayOfWeek.SATURDAY || start.getDayOfWeek() == DayOfWeek.SUNDAY){
+                numWeekend++;
+            }else{
+                numStandard++;
+            }
+        }
+        return standard.multiply(new BigDecimal(numStandard)).add(weekend.multiply(new BigDecimal(numWeekend)));
+    }
+
+    private Result<Reservation> validate(Reservation reservation){
+        Result<Reservation> result = validateNulls(reservation);
+        if (!result.isSuccess()){
+            return result;
+        }
+        validateDates(reservation,result);
+        return result;
+    }
+
+    private Result<Reservation> validateNulls(Reservation reservation){
+        Result<Reservation> result = new Result<>();
+
+        if (reservation == null){
+            result.addErrorMessage("Nothing to save.");
+            return result;
+        }
+        if (guestRepository.findById(reservation.getGuestId()) == null){
+            result.addErrorMessage("Guest is required.");
+        }
+        if (hostRepository.findById(reservation.getHostId()) == null){
+            result.addErrorMessage("Host is required.");
+        }
+        if (reservation.getStartDate() == null){
+            result.addErrorMessage("Start date is required.");
+        }
+        if (reservation.getEndDate() == null){
+            result.addErrorMessage("End date is required.");
+        }
+        return result;
+    }
+
+    private void validateDates(Reservation reservation, Result<Reservation> result){
```

Comment:
I think there is a small issue here. These two date ranges should not overlap: 2/10/2022-2/11/2022 and 2/11/2022-2/17/2022. A new reservation should be able to start on the same day that a previous reservation ends.